### PR TITLE
#5044: Add optional output to addalpha

### PR DESCRIPTION
**Pull request by KalaivaniMCW**
Add optional output tensor to addalpha
cq_id to be added once support for add, mul_unary is merged.
CI: https://github.com/tenstorrent/tt-metal/actions/runs/9442805349

### CODE REVIEW

**Review by TT-BrianLiu — APPROVED** [review]

**Comment by KalaivaniMCW** [comment]
Hi @dongjin-na ,  Could you review this PR ?

File: tt_eager/tt_dnn/op_library/composite/composite_ops.cpp
Comment (dongjin-na) [inline tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:734]:
```suggestion
    std::optional<Tensor> output_tensor,
    const MemoryConfig& output_mem_config) {
    if (output_tensor.has_value()) {
        output_tensor = add(mul_unary(input_b, alpha, output_mem_config), input_a, output_tensor, output_mem_config);
```

The output_tensor object provided by the user should be used and remain active. Using *(*output_tensor) will remove the existing output_tensor. The output_tensor variable will be replaced by the newly created tensor object by the add function.

Comment (KalaivaniMCW) [reply tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:734]: @dongjin-na  passing output_tensor to the add op is what I would ideally like to do. 
`add(mul_unary(input_b, alpha, output_mem_config), input_a, std::nullopt, output_mem_config, output_tensor);`
Until #8394 is merged both composite and backward ops implementation will have to be put on hold.

Comment (dongjin-na) [reply tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:734]: Okay. If that's what you have planned, I think it would be better to use std::optional<Tensor> and return a new Tensor object instead of using std::optional<Tensor*>. It might be better to stick with the existing approach.

What I mean is that we should keep using std::optional<Tensor> output_tensor as a parameter before #8394 is merged. Once #8394 is integrated, we can make the necessary adjustments, focusing only on the function's internal implementation without changing the interface. This approach seems more efficient to me.

Anyway, it seems like we are on the same page regarding the final task, so I'm fine with either implementation.

Comment (KalaivaniMCW) [reply tt_eager/tt_dnn/op_library/composite/composite_ops.cpp]: Hi @dongjin-na and @eyonland , 

`t3 = ttl.tensor.addalpha(t0, t1, alpha, output_tensor=t2, output_mem_config=output_mem_config)`

Using `std::optional<Tensor> output_tensor` in composite structure passes the tensor by value and the result doesn't reflect in output_tensor(t2) after the op is executed. Even though output_tensor is used for computation, we can only return the result as a new Tensor object(t3). Are you okay with this approach ? 
image (50)

I tried using reference_wrapper and was able to do modify output_tensor in memory. But I remember something about not using std::reference_wrapper, (Eyon mentioned something) not sure why but here are the results though. Whichever approach you are ok with, we'll go with that.

Add optional output tensor to addalpha op using `std::optional<std::reference_wrapper<Tensor>> output_tensor`
- The passed output_tensor is modified inside the op with the expected result. It is used for computation without creating extra tensors.

image

- Using binary op with output_tensor
Screenshot 2024-06-02 at 3 45 28 AM

File: tt_eager/tt_dnn/op_library/composite/composite_ops.cpp
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:736]: You need to pass output_tensor as the optional output tensor argument to add for this to work

**Review by tt-aho — CHANGES_REQUESTED** [review]

File: tt_eager/tt_dnn/op_library/composite/composite_ops.cpp
Comment (KalaivaniMCW) [reply tt_eager/tt_dnn/op_library/composite/composite_ops.cpp]: Yes. now that #8394 is merged I can pass the output_tensor.

File: tt_eager/tt_dnn/op_library/composite/composite_ops.cpp
Comment (dongjin-na) [inline tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:711]: Looks good to me.

Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:711]: Ideally the output_mem_config arg should be made an optional for binary as well so that we can just pass std::nullopt, but since it isn't you can pass `operation::DEFAULT_OUTPUT_MEMORY_CONFIG` instead of querying the output tensor's mem config since querying it is more expensive and the value is not needed

Comment (tt-aho) [reply tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:711]: We should file an issue to make it optional if we don't have an open issue already

Comment (KalaivaniMCW) [reply tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:711]: I will be doing it as part of this issue https://github.com/tenstorrent/tt-metal/issues/8962

**Review by tt-aho — APPROVED** [review]

**Pull request merged**
